File: BancoCoreSOAP/src/main/java/com/example/BancoCoreSoap/service/ContaService.java
package com.example.BancoCoreSoap.service;

import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.time.ZoneId;

import org.springframework.beans.factory.annotation.Autowired;
import com.example.BancoCoreSoap.repository.ClienteRepository;
import com.example.BancoCoreSoap.repository.ContaRepository;
import com.example.BancoCoreSoap.repository.TransacaoRepository;

import org.springframework.transaction.annotation.Transactional;
import com.example.demo.banco.gen.CriarContaResponse;
import com.example.demo.banco.gen.RealizarTransferenciaTEDRequest;
import com.example.demo.banco.gen.RealizarTransferenciaTEDResponse;
import com.example.demo.banco.gen.CriarContaRequest;
import com.example.BancoCoreSoap.domain.Cliente;
import com.example.BancoCoreSoap.domain.Conta;
import com.example.BancoCoreSoap.domain.TipoTransacao;
import com.example.BancoCoreSoap.domain.Transacao;

@Service
public class ContaService {
	
	@Autowired
	private ClienteRepository clienteRepository;
	
	@Autowired
	private ContaRepository contaRepository;
	
	@Autowired
	private TransacaoRepository transacaoRepository;
	
	@Transactional
	public CriarContaResponse create(CriarContaRequest request) {
		 Cliente clienteDono = clienteRepository.findById(request.getClienteId())
				 .orElseThrow(() -> new RuntimeException("Cliente não encontrado!"));
		 
		 Conta newConta = new Conta();
		 newConta.setNumeroConta(request.getNumeroConta());
		 newConta.setSaldo(request.getSaldoInicial()); 
		 newConta.setCliente(clienteDono);
		 
		 Conta savedConta = contaRepository.save(newConta);
		 
		 CriarContaResponse response = new CriarContaResponse();
		 response.setNumeroConta(savedConta.getNumeroConta());
		 response.setSaldo(savedConta.getSaldo());
		 response.setNomeCliente(savedConta.getCliente().getNome());
		 
		 return response;
	}
	
	@Transactional
	public RealizarTransferenciaTEDResponse realizarTransferencia(RealizarTransferenciaTEDRequest request) {

		Conta contaOrigem = contaRepository.findByNumeroConta(request.getContaOrigem().trim())
		        .orElseThrow(() -> new RuntimeException("Conta de origem não encontrada"));

		Conta contaDestino = contaRepository.findByNumeroConta(request.getContaDestino().trim())
		        .orElseThrow(() -> new RuntimeException("Conta de destino não encontrada"));
		
		if (contaOrigem.getSaldo().compareTo(request.getValor()) < 0) {
			throw new RuntimeException("Saldo insuficiente");
		}
		
		if (contaOrigem.getNumeroConta() == contaDestino.getNumeroConta()) {
			throw new RuntimeException("Não pode realizar uma transferencia para a mesma conta");
		}
		
		BigDecimal novoSaldoOrigem = contaOrigem.getSaldo().subtract(request.getValor());
		contaOrigem.setSaldo(novoSaldoOrigem);
		
		BigDecimal novoSaldoDestino = contaDestino.getSaldo().add(request.getValor());
		contaDestino.setSaldo(novoSaldoDestino);
		
		LocalDateTime agoraBrasilia = LocalDateTime.now(ZoneId.of("America/Sao_Paulo"));
		
		Transacao debito = new Transacao();
		debito.setConta(contaOrigem);
		debito.setDataHora(agoraBrasilia);
		debito.setTipo(TipoTransacao.TED);
		debito.setValor(request.getValor().negate());
		
		Transacao credito = new Transacao();
		credito.setConta(contaDestino);
		credito.setDataHora(agoraBrasilia);
		credito.setTipo(TipoTransacao.TED);
		credito.setValor(request.getValor());
		
		Transacao savedDebito = transacaoRepository.save(debito);
		Transacao savedCredito = transacaoRepository.save(credito);
		
		RealizarTransferenciaTEDResponse response = new RealizarTransferenciaTEDResponse();
		response.setStatus("SUCESSO");
		response.setMensagem("Transferencia TED realizada com sucesso.");
		response.setNovoSaldoDestino(novoSaldoDestino);
		response.setNovoSaldoOrigem(novoSaldoOrigem);
		response.setTransacaoCreditoId(savedCredito.getId());
		response.setTransacaoDebitoId(savedDebito.getId());
		
		return response;
	}
}

File: BancoRestApi/src/main/kotlin/org/example/bancorestapi/service/ContaService.kt
package org.example.bancorestapi.service

import org.springframework.transaction.annotation.Transactional
import org.example.bancorestapi.domain.ChavePix
import org.example.bancorestapi.dto.PixTransferRequestDTO
import org.example.bancorestapi.dto.PixTransferResponseDTO
import org.example.bancorestapi.domain.Conta
import org.example.bancorestapi.domain.TipoTransacao
import org.example.bancorestapi.domain.Transacao
import org.example.bancorestapi.domain.toResponseDTO
import org.example.bancorestapi.dto.ChavePixResponseDTO
import org.example.bancorestapi.repository.ChavePixRepository
import org.example.bancorestapi.repository.ContaRepository
import org.example.bancorestapi.repository.TransacaoRepository
import org.springframework.stereotype.Service
import java.math.BigDecimal
import java.time.LocalDateTime
import java.time.ZoneId

@Service
class ContaService (
    private val transacaoRepository: TransacaoRepository,
    private val chavePixRepository: ChavePixRepository,
    private val contaRepository: ContaRepository
) {
    @Transactional
    fun realizarTransferenciaPix(request: PixTransferRequestDTO): PixTransferResponseDTO {

        if (request.valor <= BigDecimal.ZERO) {
            throw RuntimeException("O valor da transferência deve ser positivo.")
        }

        val contaOrigem: Conta = contaRepository.findByNumeroConta(request.numeroContaOrigem.trim())
            ?: throw RuntimeException("Conta de origem não encontrada")

        val chavePix: ChavePix = chavePixRepository.findByChave(request.chaveDestino)
            ?: throw RuntimeException("Chave pix não encontrada")

        val contaDestinoDaChave: Conta = chavePix.conta

        val idDestino = contaDestinoDaChave.id
            ?: throw RuntimeException("A chave PIX não tem ID de conta associado")

        val contaDestino: Conta = contaRepository.findById(idDestino)
            ?: throw RuntimeException("A conta de destino (ID: $idDestino) não foi encontrada.")

        if (contaOrigem.saldo < request.valor) {
            throw RuntimeException("Saldo indisponivel")
        }

        if (contaOrigem.numeroConta == contaDestino.numeroConta) {
            throw RuntimeException("Não pode realizer uma transferencia para a mesma conta")
        }

        contaOrigem.saldo = contaOrigem.saldo.subtract(request.valor)
        contaDestino.saldo = contaDestino.saldo.add(request.valor)

        contaRepository.saveAll(listOf(contaOrigem, contaDestino))

        val agoraBrasilia: LocalDateTime = LocalDateTime.now(ZoneId.of("America/Sao_Paulo"))

        val debito = Transacao()
        debito.conta = contaOrigem
        debito.dataHora = agoraBrasilia
        debito.tipo = TipoTransacao.PIX
        debito.valor = request.valor.negate()

        val credito = Transacao()
        credito.conta = contaDestino
        credito.dataHora = agoraBrasilia
        credito.tipo = TipoTransacao.PIX
        credito.valor = request.valor

        transacaoRepository.saveAll(listOf(debito, credito))

        val response = PixTransferResponseDTO(
            status = "Sucesso",
            mensagem = "Transferencia PIX realizada com sucesso",
            novoSaldoOrigem = contaOrigem.saldo
        )

        return response
    }

    @Transactional
    fun getChavePix(numeroConta: String) : List<ChavePixResponseDTO> {
        val conta: Conta = contaRepository.findByNumeroConta(numeroConta)
            ?: throw RuntimeException("Esse numero não corresponde a nenhuma conta")

        val chavesPixEncontradas: List<ChavePix> = chavePixRepository.findByConta(conta)

        return chavesPixEncontradas.map { chavePix ->
            chavePix.toResponseDTO()
        }

    }
}